File: src/util_math.py
# Copyright 2021 MIT Probabilistic Computing Project
# Apache License, Version 2.0, refer to LICENSE.txt

import math
import random

inf = float('inf')

def linspace(start, stop, num=50, endpoint=True):
    """linspace from a to b with n entries."""
    step = (stop - start) / (num - endpoint)
    return [start + step*i for i in range(num)]

def log_linspace(a, b, n):
    """linspace from a to b with n entries over log scale."""
    points = linspace(math.log(a), math.log(b), num=n)
    return [math.exp(x) for x in points]

def log_normalize(log_weights):
    """Return log of the sum of exponentials of input list divided by sum."""
    Z = logsumexp(log_weights)
    return [x - Z for x in log_weights]

def log_choices(population, log_weights, k=1, prng=None):
    """Draw from a population given a list of log probabilities."""
    log_weights_normalized = log_normalize(log_weights)
    weights = [math.exp(w) for w in log_weights_normalized]
    return (prng or random).choices(population, weights, k=k)

def logsumexp(array):
    """Return log of the sum of exponentials of input elements."""
    if len(array) == 0:
        return float('-inf')

    # m = +inf means addends are all +inf, hence so are sum and log.
    # m = -inf means addends are all zero, hence so is sum, and log is
    # -inf.  But if +inf and -inf are among the inputs, or if input is
    # NaN, let the usual computation yield a NaN.
    m = max(array)
    if math.isinf(m) \
            and min(array) != -m \
            and all(not math.isnan(a) for a in array):
        return m

    # Since m = max{a_0, a_1, ...}, it follows that a <= m for all a,
    # so a - m <= 0; hence exp(a - m) is guaranteed not to overflow.
    return m + math.log(sum(math.exp(a - m) for a in array))

def logmeanexp(array):
    """Return log of the mean of exponentials of input elements."""
    if len(array) == 0:
        return -inf

    # Treat -inf values as log 0 -- they contribute zero to the sum in
    # logsumexp, but one to the count.
    #
    # If we pass -inf values through to logsumexp, and there are also
    # +inf values, then we get NaN -- but if we had averaged exp(-inf)
    # = 0 and exp(+inf) = +inf, we would sensibly get +inf, whose log
    # is still +inf, not NaN.  So strip -inf values first.
    #
    # Can't say `a > -inf' because that excludes NaNs, but we want to
    # include them so they propagate.
    noninfs = [a for a in array if not a == -inf]

    # probs = map(exp, array)
    # log(mean(probs))
    #   = log(sum(probs) / len(probs))
    #   = log(sum(probs)) - log(len(probs))
    #   = log(sum(map(exp, array))) - log(len(array))
    #   = logsumexp(array) - log(len(array))
    return logsumexp(noninfs) - math.log(len(array))

File: cxx/util_math.cc
// Copyright 2021 MIT Probabilistic Computing Project
// Apache License, Version 2.0, refer to LICENSE.txt

#include <algorithm>

#include "util_math.hh"

using std::vector;

const double INF = std::numeric_limits<double>::infinity();

// http://matlab.izmiran.ru/help/techdoc/ref/betaln.html
double lbeta(int z, int w) {
    return lgamma(z) + lgamma(w) - lgamma(z + w);
}

vector<double> linspace(double start, double stop, int num, bool endpoint) {
    double step = (stop - start) / (num - endpoint);
    vector<double> v;
    for (int i = 0; i < num; i++) {
        v.push_back(start + step * i);
    }
    return v;
}

vector<double> log_linspace(double start, double stop, int num, bool endpoint) {
    auto v = linspace(log(start), log(stop), num, endpoint);
    for (int i = 0; i < v.size(); i++) {
        v[i] = exp(v[i]);
    }
    return v;
}

vector<double> log_normalize(const std::vector<double> &weights){
    double Z = logsumexp(weights);
    vector<double> result(weights.size());
    for (int i = 0; i < weights.size(); i++) {
        result[i] = weights[i] - Z;
    }
    return result;
}

double logsumexp(const vector<double> &weights) {
    double m = *std::max_element(weights.begin(), weights.end());
    double s = 0;
    for (auto w : weights) {
        s += exp(w - m);
    }
    return log(s) + m;
}

int choice(const std::vector<double> &weights, PRNG *prng) {
    std::discrete_distribution<int> dist(weights.begin(), weights.end());
    int idx = dist(*prng);
    return idx;
}

int log_choice(const std::vector<double> &weights, PRNG *prng) {
    vector<double> log_weights_norm = log_normalize(weights);
    vector<double> weights_norm;
    for (double w : log_weights_norm) {
        weights_norm.push_back(exp(w));
    }
    return choice(weights_norm, prng);
}

vector<vector<int>> product(const vector<vector<int>> &lists) {
    // https://rosettacode.org/wiki/Cartesian_product_of_two_or_more_lists#C.2B.2B
    vector<vector<int>> result;
    for (const auto &l : lists) {
        if (l.size() == 0) {
            return result;
        }
    }
    for (const auto &e : lists[0]) {
        result.push_back({e});
    }
    for (size_t i = 1; i < lists.size(); ++i) {
        vector<vector<int>> temp;
        for (auto &e : result) {
            for (auto f : lists[i]) {
                auto e_tmp = e;
                e_tmp.push_back(f);
                temp.push_back(e_tmp);
          }
        }
        result = temp;
      }
      return result;
    }
